### PR TITLE
[DotLiquid] Add TraceInfo for HL7 v2 conversion

### PR DESCRIPTION
**Pull request by BoyaWu10**
## Description

Add TraceInfo for HL7 v2 conversion. If (a) `-t` is specified in command-line tool or (b) `Hl7v2TraceInfo` is passed in conversion interface, TraceInfo will be provided along with conversion result. Currently, TraceInfo contains unused segments.

Some other changes in this PR:
- Refine error handling in `ConverterLogicHandler`, as the one in template management.
- Use `RenderParameters.FromContext` which is provided directly in DotLiquid.
- Make `SafeList` enumerable.
- Put `.editorconfig` in solution items, as suggested by IDE.

## Related issues

Addresses [issue #].

## Testing

Describe how this change was tested.

### CODE REVIEW

**Comment by qiwjin** [comment]
>         private static void ConvertBatchFiles(ConverterOptions options)

possible to call ConvertSingleFile in this function?

---
Refers to: src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs:74 in 3ccd30c. [](commit_id = 3ccd30cf0ba1683ef4fc19f8df0c9a876e2a8340, deletion_comment = False)

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+
+namespace Microsoft.Health.Fhir.Liquid.Converter.Hl7v2.Models
+{
+    public class UnusedHl7v2Field
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs:10]: >UnusedHl7v2Field [](start = 17, length = 16)

possible to provide the position of the field?

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/Hl7v2TraceInfo.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                    if (j == 0 && segment.Fields[j] is Hl7v2Field segmentIdField)
+                    {
+                        unusedSegment.Type = segmentIdField.Value;
+                        continue;
+                    }
+
+                    if (segment.Fields[j] is Hl7v2Field field)
+                    {
+                        var unusedField = new UnusedHl7v2Field(j);
+                        for (var k = 0; k < field.Components.Count; ++k)
```
Comment (moria97) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/Hl7v2TraceInfo.cs:49]: Maybe we can use LinQ expression here unusedField.Component = field.Components.Where(component is unused).Select(new UnusedComponent());

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs (hunk 1 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs:10]: Yes, provided in the output. One example output could be found [here](https://microsoftapc-my.sharepoint.com/:u:/g/personal/boywu_microsoft_com/EQrqXvlERhhKuECMr99KEDQBNeyn69WepGC0v-gFI6A63Q?e=RdnY8y).

File: src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs [hunk 3]
```diff
@@ -56,9 +57,10 @@ private static void ConvertSingleFile(ConverterOptions options)
                 var dataType = GetDataTypes(options.TemplateDirectory);
                 var dataProcessor = CreateDataProcessor(dataType);
                 var templateProvider = CreateTemplateProvider(dataType, options.TemplateDirectory);
-                var resultString = dataProcessor.Convert(options.InputDataContent, options.RootTemplate, templateProvider);
-                var result = new ConverterResult(ProcessStatus.OK, resultString);
-                WriteOutputFile(options.OutputDataFile, JsonConvert.SerializeObject(result, Formatting.Indented));
+                var traceInfo = CreateTraceInfo(dataType, options.IsTraceInfo);
+                var resultString = dataProcessor.Convert(options.InputDataContent, options.RootTemplate, templateProvider, traceInfo);
+                var result = new ConverterResult(ProcessStatus.OK, resultString, traceInfo);
+                WriteOutputFile(options.OutputDataFile, JsonConvert.SerializeObject(result, Formatting.Indented, new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore }));
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs]: Created a new function called `SaveConverterResult` in the update.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/Hl7v2TraceInfo.cs (hunk 2 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/Hl7v2TraceInfo.cs]: I tried using LINQ expression here and resulted in a long linq expression, as new features are added in the update (recording the index in the segment). Another `Select` needs to be added before `Where` to preserve the original index `k`.

**Review by moria97 — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Liquid.Converter.Tool/Program.cs [hunk 4]
```diff
@@ -14,19 +14,23 @@ namespace Microsoft.Health.Fhir.Liquid.Converter.Tool
 {
     public class Program
     {
-        public static async Task Main(string[] args)
+        public static async Task<int> Main(string[] args)
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Liquid.Converter.Tool/Program.cs:17]: Will keep this in sync with changes in template management.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Models/TraceInfo.cs [hunk 5]
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Health.Fhir.Liquid.Converter.Models
 {
     /// <summary>
     /// TraceInfo records processing details during conversion.
-    /// It is going to be implemented in coming releases.
+    /// For HL7 v2 conversion, please use its inherited class Hl7v2TraceInfo.
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Models/TraceInfo.cs:10]: also consider ccda case if any shared field or method.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs (hunk 1 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Field.cs]: Simplified output TraceInfo in above link.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Fhir.Liquid.Converter.Hl7v2.Models
+{
+    public class UnusedHl7v2Component
+    {
+        public UnusedHl7v2Component(int index, string value)
+        {
+            Index = index;
```
Comment (yankunhuang-pku) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs]: The end position is the index plus the length of value, right?

File: src/Microsoft.Health.Fhir.Liquid.Converter/Models/TraceInfo.cs (hunk 5 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Models/TraceInfo.cs:10]: Yes, would it be better if we simplify the output to "UnusedLines" and "UnusedIntervals" and make them shared for both HL7 v2 and CCDA?

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs (hunk 6 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs]: The source code is updated. The start position is `Offset` and the end position is `Offset + Value.Length`.

**Review by yankunhuang-pku — APPROVED** [review]

**Review by ACMoretxj — APPROVED** [review]

**Review by sowu880 — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs (hunk 6 above)
Comment (qiwjin) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs]: I feel you can provider readonly properties for start and end positions 

---
In reply to: [540747689](https://github.com/microsoft/FHIR-Converter/pull/123#discussion_r540747689) [](ancestors = 540747689)

**Review by qiwjin — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs (hunk 6 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/UnusedHl7v2Component.cs]: Thanks, added end index in the output.

**Pull request merged**